app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx

```diff
@@ -1,48 +1,155 @@
-  }
-
-  private mapChanged = (key: string, values: { [key: string]: string }) => {
-    this.props.updateStageField({ [key]: this.expand(values) });
-  };
-
-  public render() {
-    const {
-      stage: { variables = [] },
-      pipeline,
-    } = this.props;
-
-    // Flattens an array of objects {key, value} into a single object with the respective keys/values
-    const variablesObject = variables.reduce(
-      // tslint:disable-next-line:prefer-object-spread
-      (acc: any, { key, value }: IEvaluatedVariable) => Object.assign(acc, { [key]: value }),
-      {},
-    );
-
-    return (
-      <div className="form-horizontal">
-        <StageConfigField label="Variables to evaluate">
-          <MapEditor
-            model={variablesObject}
-            allowEmpty={true}
-            onChange={(v: any) => this.mapChanged('variables', v)}
-            pipeline={pipeline}
-            valueCanContainSpel={true}
-          />
-        </StageConfigField>
-      </div>
-    );
-  }
+const variableNameValidator: IValidator = (val: string, label: string) =>
+  !val.match(/^[a-zA-Z0-9_]+$/) && errorMessage(`${label} should consist only of letters, numbers, or underscore`);
+
+const getDuplicateKeyValidator = (variables: IEvaluatedVariable[]) => {
+  const keyCounts = countBy(variables.map(x => x.key), x => x);
+  return (key: string) => keyCounts[key] > 1 && `Duplicate key '${key}'`;
+};
+
+export function validateEvaluateVariablesStage(stage: IStage) {
+  const formValidator = new FormValidator(stage);
+  const duplicateKeyValidator = getDuplicateKeyValidator(stage.variables);
+  formValidator.field('variables').withValidators(
+    formValidator.arrayForEach(item => {
+      item
+        .field('key', 'Variable Name')
+        .required()
+        .withValidators(variableNameValidator, duplicateKeyValidator);
+      item.field('value', 'Expression').required();
+    }),
+  );
+  return formValidator.validateForm();
+}
+
+export function EvaluateVariablesStageConfig(props: IStageConfigProps) {
+  const { application, stage, pipeline, updateStage } = props;
+  return (
+    <FormikStageConfig
+      application={application}
+      stage={stage}
+      pipeline={pipeline}
+      validate={validateEvaluateVariablesStage}
+      onChange={updateStage}
+      render={renderProps => {
+        return (
+          <LayoutProvider value={StandardFieldLayout}>
+            <EvaluateVariablesStageForm {...renderProps} />
+          </LayoutProvider>
+        );
+      }}
+    />
+  );
+}
+
+function EvaluateVariablesStageForm(props: IFormikStageConfigInjectedProps) {
+  const { formik } = props;
+
+  React.useEffect(() => {
+    const { variables = [] } = props.formik.values;
+    const initialTouched = variables.map(() => ({ key: true, value: true }));
+    formik.setTouched({ variables: initialTouched } as any);
+  }, []);
+
+  const [deleteCount, setDeleteCount] = React.useState(0);
```

Comment:
Would it be possible to generate and persist some UUID for each variable so we don't need this? :pray:

Reply:
theoretically we don't even need to persist it, as long as it's stable per-mount.